File: python/src/revenant/ui/workflows.py
"""Shared signing and verification workflows.

UI-agnostic orchestration of signing and verification operations.
CLI and GUI are thin wrappers around these functions.

Constraints:
- No stdout/stderr output (no print)
- No sys.exit()
- No tkinter imports
- No argparse imports
- No threading (caller's responsibility)
- Returns structured results, never raises on business errors
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import TYPE_CHECKING

from ..config import (
    get_active_profile,
    get_signer_info,
    register_active_profile_tls,
)
from ..constants import DEFAULT_POSITION
from ..core.appearance import extract_cert_fields, extract_display_fields
from ..errors import AuthError, RevenantError, TLSError
from .helpers import atomic_write

if TYPE_CHECKING:
    from pathlib import Path

    from ..core.pdf import VerificationResult

_logger = logging.getLogger(__name__)


# ── Result types ──────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class SigningResult:
    """Result of a single signing operation."""

    ok: bool
    auth_failed: bool = False
    tls_error: bool = False
    error_message: str | None = None
    output_path: Path | None = None
    output_size: int = 0


@dataclass(frozen=True, slots=True)
class VerifyEntry:
    """Structured verification data for a single signature."""

    index: int
    total: int
    valid: bool
    signer_name: str
    detail_lines: list[str]


@dataclass(frozen=True, slots=True)
class VerifyResult:
    """Structured verification results for all signatures in a PDF."""

    all_valid: bool
    total_count: int
    failed_count: int
    entries: list[VerifyEntry]


# ── Error classification ──────────────────────────────────────────


def _classify_error(error: Exception) -> SigningResult:
    """Convert a caught exception into a SigningResult."""
    if isinstance(error, AuthError):
        return SigningResult(ok=False, auth_failed=True, error_message=str(error))

    if isinstance(error, TLSError):
        return SigningResult(ok=False, tls_error=True, error_message=str(error))

    if isinstance(error, (RevenantError, ValueError)):
        return SigningResult(ok=False, error_message=str(error))

    _logger.exception("Unexpected error during signing")
    return SigningResult(
        ok=False,
        error_message="An unexpected error occurred. Check logs for details.",
    )


# ── Signing workflows ────────────────────────────────────────────


def sign_one_embedded(
    pdf_bytes: bytes,
    output_path: Path,
    url: str,
    username: str,
    password: str,
    timeout: int,
    *,
    name: str | None = None,
    position: str = DEFAULT_POSITION,
    page: int | str = "last",
    image_path: str | None = None,
    visible: bool = True,
    font: str | None = None,
    reason: str = "Signed with Revenant",
    fields: list[str] | None = None,
) -> SigningResult:
    """Orchestrate embedded PDF signing.

    Creates transport, calls core signing, writes output atomically.
    Never raises on business errors -- all captured in the result.

    Args:
        pdf_bytes: Raw PDF file content (caller reads the file).
        output_path: Where to write the signed PDF.
        url: SOAP endpoint URL.
        username: Authentication username.
        password: Authentication password.
        timeout: Request timeout in seconds.
        name: Signer display name for the signature field.
        position: Signature position preset.
        page: Page for signature (0-based int, "first", or "last").
        image_path: Optional path to signature image.
        visible: If False, create invisible signature.
        font: Font registry key.
        reason: Signature reason string.
        fields: Pre-extracted display fields for the appearance.

    Returns:
        SigningResult with outcome, output path, and size.
    """
    try:
        from ..core.signing import sign_pdf_embedded
        from ..network import SoapSigningTransport

        register_active_profile_tls()
        transport = SoapSigningTransport(url)
        signed_pdf = sign_pdf_embedded(
            pdf_bytes,
            transport,
            username,
            password,
            timeout,
            name=name,
            position=position,
            page=page,
            reason=reason,
            image_path=image_path,
            fields=fields,
            visible=visible,
            font=font,
        )
    except Exception as e:
        return _classify_error(e)

    try:
        atomic_write(output_path, signed_pdf)
    except PermissionError:
        return SigningResult(
            ok=False,
            error_message=f"Permission denied: {output_path}",
        )

    return SigningResult(
        ok=True,
        output_path=output_path,
        output_size=len(signed_pdf),
    )


def sign_one_detached(
    pdf_bytes: bytes,
    output_path: Path,
    url: str,
    username: str,
    password: str,
    timeout: int,
) -> SigningResult:
    """Orchestrate detached PDF signing.

    Creates transport, calls core signing, writes output atomically.
    Never raises on business errors -- all captured in the result.

    Args:
        pdf_bytes: Raw PDF file content (caller reads the file).
        output_path: Where to write the .p7s signature.
        url: SOAP endpoint URL.
        username: Authentication username.
        password: Authentication password.
        timeout: Request timeout in seconds.

    Returns:
        SigningResult with outcome, output path, and size.
    """
    try:
        from ..core.signing import sign_pdf_detached
        from ..network import SoapSigningTransport

        register_active_profile_tls()
        transport = SoapSigningTransport(url)
        cms_signature = sign_pdf_detached(pdf_bytes, transport, username, password, timeout)
    except Exception as e:
        return _classify_error(e)

    try:
        atomic_write(output_path, cms_signature)
    except PermissionError:
        return SigningResult(
            ok=False,
            error_message=f"Permission denied: {output_path}",
        )

    return SigningResult(
        ok=True,
        output_path=output_path,
        output_size=len(cms_signature),
    )


# ── Verification workflow ─────────────────────────────────────────


def format_verify_results(results: list[VerificationResult]) -> VerifyResult:
    """Convert raw verification results into structured display data.

    Args:
        results: Raw verification results from verify_all_embedded_signatures.

    Returns:
        VerifyResult with structured entries for UI display.
    """
    total = len(results)
    entries: list[VerifyEntry] = []
    failed = 0

    for i, result in enumerate(results):
        signer = result.get("signer")
        signer_name = (signer.get("name") or "Unknown") if signer else "Unknown"
        valid = result["valid"]
        if not valid:
            failed += 1

        detail_lines: list[str] = []
        for detail in result["details"]:
            detail_lines.extend(detail.split("\n"))

        entries.append(
            VerifyEntry(
                index=i,
                total=total,
                valid=valid,
                signer_name=signer_name,
                detail_lines=detail_lines,
            )
        )

    return VerifyResult(
        all_valid=(failed == 0),
        total_count=total,
        failed_count=failed,
        entries=entries,
    )


# ── Field extraction helper ──────────────────────────────────────


def resolve_sig_fields() -> list[str] | None:
    """Extract display fields from the active profile's sig_fields config.

    Returns:
        Ordered list of display strings, or None if no sig_fields configured.
    """
    profile = get_active_profile()
    if not profile or not profile.sig_fields:
        return None
    signer_info = get_signer_info()
    cert_values = extract_cert_fields(profile.cert_fields, signer_info)
    return extract_display_fields(profile.sig_fields, cert_values)

File: typescript/src/cli/workflows.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Shared signing and verification workflows.
 *
 * UI-agnostic orchestration of signing and verification operations.
 * CLI is a thin wrapper around these functions.
 *
 * Constraints:
 * - No stdout/stderr output (no console.log)
 * - No process.exit()
 * - Returns structured results, never raises on business errors
 */

import { getActiveProfile, getSignerInfo, registerActiveProfileTls } from "../config/index.js";
import { DEFAULT_POSITION } from "../constants.js";
import { extractCertFields, extractDisplayFields } from "../core/appearance/index.js";
import type { VerificationResult } from "../core/pdf/index.js";
import { AuthError, isNodeError, RevenantError, TLSError } from "../errors.js";
import { atomicWrite } from "./helpers.js";

// -- Result types -------------------------------------------------------------

export interface SigningResult {
  ok: boolean;
  authFailed: boolean;
  tlsError: boolean;
  errorMessage: string | null;
  outputPath: string | null;
  outputSize: number;
}

export interface VerifyEntry {
  index: number;
  total: number;
  valid: boolean;
  signerName: string;
  detailLines: string[];
}

export interface VerifyResult {
  allValid: boolean;
  totalCount: number;
  failedCount: number;
  entries: VerifyEntry[];
}

// -- Error classification -----------------------------------------------------

function classifyError(error: unknown): SigningResult {
  if (error instanceof AuthError) {
    return {
      ok: false,
      authFailed: true,
      tlsError: false,
      errorMessage: error.message,
      outputPath: null,
      outputSize: 0,
    };
  }

  if (error instanceof TLSError) {
    return {
      ok: false,
      authFailed: false,
      tlsError: true,
      errorMessage: error.message,
      outputPath: null,
      outputSize: 0,
    };
  }

  if (error instanceof RevenantError || error instanceof Error) {
    return {
      ok: false,
      authFailed: false,
      tlsError: false,
      errorMessage: error.message,
      outputPath: null,
      outputSize: 0,
    };
  }

  return {
    ok: false,
    authFailed: false,
    tlsError: false,
    errorMessage: "An unexpected error occurred.",
    outputPath: null,
    outputSize: 0,
  };
}

// -- Signing workflows --------------------------------------------------------

export interface EmbeddedSignOptions {
  name?: string | null;
  position?: string;
  page?: number | string;
  imagePath?: string | null;
  visible?: boolean;
  font?: string | null;
  reason?: string;
  fields?: string[] | null;
}

/**
 * Orchestrate embedded PDF signing.
 *
 * Creates transport, calls core signing, writes output atomically.
 * Never raises on business errors -- all captured in the result.
 */
export async function signOneEmbedded(
  pdfBytes: Uint8Array,
  outputPath: string,
  url: string,
  username: string,
  password: string,
  timeout: number,
  options: EmbeddedSignOptions = {},
): Promise<SigningResult> {
  let signedPdf: Uint8Array;
  try {
    const { signPdfEmbedded } = await import("../core/signing.js");
    const { SoapSigningTransport } = await import("../network/soap-transport.js");

    await registerActiveProfileTls();
    const transport = new SoapSigningTransport(url);
    signedPdf = await signPdfEmbedded(pdfBytes, transport, username, password, timeout, {
      name: options.name,
      position: options.position ?? DEFAULT_POSITION,
      page: options.page ?? "last",
      reason: options.reason ?? "",
      imagePath: options.imagePath,
      fields: options.fields,
      visible: options.visible ?? true,
      font: options.font,
    });
  } catch (e) {
    return classifyError(e);
  }

  try {
    atomicWrite(outputPath, signedPdf);
  } catch (e) {
    if (isNodeError(e, "EACCES")) {
      return {
        ok: false,
        authFailed: false,
        tlsError: false,
        errorMessage: `Permission denied: ${outputPath}`,
        outputPath: null,
        outputSize: 0,
      };
    }
    throw e;
  }

  return {
    ok: true,
    authFailed: false,
    tlsError: false,
    errorMessage: null,
    outputPath,
    outputSize: signedPdf.length,
  };
}

/**
 * Orchestrate detached PDF signing.
 */
export async function signOneDetached(
  pdfBytes: Uint8Array,
  outputPath: string,
  url: string,
  username: string,
  password: string,
  timeout: number,
): Promise<SigningResult> {
  let cmsSignature: Uint8Array;
  try {
    const { signPdfDetached } = await import("../core/signing.js");
    const { SoapSigningTransport } = await import("../network/soap-transport.js");

    await registerActiveProfileTls();
    const transport = new SoapSigningTransport(url);
    cmsSignature = await signPdfDetached(pdfBytes, transport, username, password, timeout);
  } catch (e) {
    return classifyError(e);
  }

  try {
    atomicWrite(outputPath, cmsSignature);
  } catch (e) {
    if (isNodeError(e, "EACCES")) {
      return {
        ok: false,
        authFailed: false,
        tlsError: false,
        errorMessage: `Permission denied: ${outputPath}`,
        outputPath: null,
        outputSize: 0,
      };
    }
    throw e;
  }

  return {
    ok: true,
    authFailed: false,
    tlsError: false,
    errorMessage: null,
    outputPath,
    outputSize: cmsSignature.length,
  };
}

// -- Verification workflow ----------------------------------------------------

/**
 * Convert raw verification results into structured display data.
 */
export function formatVerifyResults(results: VerificationResult[]): VerifyResult {
  const total = results.length;
  const entries: VerifyEntry[] = [];
  let failed = 0;

  for (let i = 0; i < results.length; i++) {
    const result = results[i];
    if (result === undefined) continue;
    const signer = result.signer;
    const signerName = signer?.name ?? "Unknown";
    const valid = result.valid;
    if (!valid) failed++;

    const detailLines: string[] = [];
    for (const detail of result.details) {
      detailLines.push(...detail.split("\n"));
    }

    entries.push({
      index: i,
      total,
      valid,
      signerName,
      detailLines,
    });
  }

  return {
    allValid: failed === 0,
    totalCount: total,
    failedCount: failed,
    entries,
  };
}

// -- Field extraction helper --------------------------------------------------

/**
 * Extract display fields from the active profile's sig_fields config.
 */
export function resolveSigFields(): string[] | null {
  const profile = getActiveProfile();
  if (!profile || profile.sigFields.length === 0) {
    return null;
  }
  const signerInfo = getSignerInfo();
  const certValues = extractCertFields(profile.certFields, {
    ...signerInfo,
  });
  return extractDisplayFields(profile.sigFields, certValues);
}
